students/rmart300/lesson05/except_exercise.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/python
+
+"""
+An exercise in playing with Exceptions.
+Make lots of try/except blocks for fun and profit.
+
+Make sure to catch specifically the error you find, rather than all errors.
+"""
+
+from except_test import fun, more_fun, last_fun
+
+log = open('log.txt','w')
+# Figure out what the exception is, catch it and while still
+# in that catch block, try again with the second item in the list
+first_try = ['spam', 'cheese', 'mr death']
+
+try:
+    joke = fun(first_try[0])
+except NameError as e:
+    log.write('caught the name error' + str(e))
+    joke = fun(first_try[1])
+
+# Here is a try/except block. Add an else that prints not_joke
+try:
+    not_joke = fun(first_try[2])
+except SyntaxError:
+    print('Run Away!')
+else:
+    print(not_joke)
+
+# What did that do? You can think of else in this context, as well as in
+# loops as meaning: "else if nothing went wrong"
+# (no breaks in  loops, no exceptions in try blocks)
+
+# Figure out what the exception is, catch it and in that same block
+#
+# try calling the more_fun function with the 2nd language in the list,
+# again assigning it to next_joke.
+#
+# If there are no exceptions, call the more_fun function with the last
+# language in the list regardless of whether there was an exception
+
+# Finally, while still in the try/except block and regardless of whether
+# there were any exceptions, call the function last_fun with no
+# parameters. (pun intended)
+
+langs = ['java', 'c', 'python']
+
+try:
+   more_joke = more_fun(langs[0])
+except IndexError as e:
+    log.write('caught the index error' + str(e))
+    try:
+        next_joke = more_fun(langs[1])
+        #print(next_joke)
+    except Exception:
+        log.write(e)
```

Comment:
It's generally a better idea to catch specific exceptions that you know might occur and that you know how to handle rather than catching the base Exception object. The danger in catching the base object (or using bare `except`) is that there are so many different unknown exceptions/errors that could be slipping by silently and make it harder to debug. 

But I believe what you're trying to do here is log Exception in order to identify any unanticipated errors, and that's a valid approach as well, but this should be:

```
    except Exception as e:  
        log.write(e) 
```
FYI the [logging](https://docs.python.org/3/library/logging.html#logging.Logger.exception) module in the Python Standard Library provides an `exception` method for just this purpose.